Step 3

imcool2551 opened the pull request:
3단계 구현을 완료하고 리뷰요청 드립니다!

약간 아쉬운 부분이 있습니다. 인터페이스를 이용한 다형성을통해 최대한 if 문을 없애보려 했지만 Frame 구현체의 `canGetScore()`, `getScore()` 메서드에서는 if문을 통한 인스턴스의 타입 체크를 없애는데 실패했습니다 ㅠㅠ.

프레임의 결과(Miss/Spare/Strike)와 상관없이 다음 프레임의 spareBonus, strikeBonus 를 모두 `getScoreWithBonus()` 에 매개변수로 전달하고 내부에서 미스, 스페어, 스트라이크에 따라 점수를 계산하면 타입 체크를 피할 수 있지만, NormalFrame은 다음 프레임이 세팅이 안 되어 있는 경우 보너스 점수를 계산하려하면 NPE가 발생하고 FinalFrame은 투구가 다 끝나지 않았을 경우 보너스 점수를 계산하려하면 FinalPinNumbers에서 IndexOutOfBounds 예외가 발생해서 어쩔 수 없이 프레임의 결과에 따른 분기처리를 했습니다.

imcool2551 committed:
apply-feedback: README

상세한 클래스 설계 대신 기능 목록 위주로 정리했습니다.

imcool2551 committed:
apply-feedback: PinStatus, FrameResult

프레임의 결과를 enum PinStatus 대신, FrameResult 인터페이스와 구현 클래스들로 표현했습니다.

imcool2551 committed:
apply-feedback: PinStatus 대신 FrameResult 사용

imcool2551 committed:
apply-feedback: getter 제거

imcool2551 committed:
apply-feedback

1. Collections 클래스의 유틸 메서드 사용
2. Bowling 에서 Player 필드를 사용하지 않기 때문에 제거 -> 래핑하는 클래스가 Frames 뿐이므로 개념상 Frames 클래스와 동일 -> Bowling 클래스 제거

imcool2551 committed:
feat+test: FrameResult, Miss, Spare, Strike

각 프레임의 결과를 나타내는 Miss, Spare, Strike 클래스를 추가했습니다.
세 클래스의 공통 인터페이스로 FrameResult를 추가했습니다. 점수를 계산하는 추상 메서드 하나가 있습니다.
Spare, Strike 는 매개변수로 넘어온 보너스 점수에 10을 더해서 반환하고, Miss는 보너스에 상관 없이 생성할 때 전달한 두 정수의 합을 반환합니다.

imcool2551 committed:
refactor: PinNo

Miss 생성할 때 두 정수 생성자로 전달.
Test에서 반복된 검증 @MethodSource를 통해 중복 제거

imcool2551 committed:
refactor: PinNumbers

기존에 NormalFrame에서만 사용하던 PinNumbers를 인터페이스로 만들고 NormalFrame, FinalFrame에서 사용할 NormalPinNumbers, FinalPinNumbers를 따로 정의했습니다.
1. isFull(): 핀이 꽉찼는지 검사
2. addPin(int pinNo): 핀 추가
3. toExpression(): 문자열 표현 반환
4. getResult(): 핀 번호로 프레임 결과 반환 (Miss/Spare/Strike)
5. spareBonus(): 이전 프레임이 스페어일 경우 합산할 보너스 점수 반환
6. strikeBonus(): 이전 프레임이 스트라이크일 경우 합산할 보너스 점수 반환

imcool2551 committed:
feat+test: FinalPinNumbers

PinNumbers 인터페이스의 구현체중 하나로 FinalFrame의 투구 결과를 표현합니다.
특이사항으로, FinalFrame은 추가 점수를 다음 프레임의 PinNumbers가 아닌 자신의 PinNumbers를 통해 계산하기 때문에
getOwnSpareBonus(), getOwnStrikeBonus() 를 추가로 구현했습니다.

imcool2551 committed:
feat+test: NormalPinNumbers

1~9번 프레임 NormalFrame의 투구 결과를 표현하는 클래스입니다. FinalPinNumbers처럼 PinNumbers 인터페이스를 구현했습니다.

imcool2551 committed:
feat: Frame

NormalFrame, FinalFrame이 구현할 인터페이스입니다.

1. getNextFrame(int pinNo): 다음 프레임을 추가합니다.
2. toExpression(): 문자열 표현을 반환합니다.
3. canGetScore(): 현재 점수 합산이 가능한지 반환합니다. Miss, Spare, Strike 여부에 따라 반환값이 달라집니다.
4. getScore(): 점수를 합산하여 반환합니다.
5. spareBonusForPreviousFrame(): 이전 프레임이 스페어인 경우를 위해 스페어 추가 점수를 반환합니다.
6. strikeBonusForPreviousFrame(): 이전 프레임이 스트라이크인 경우를 위해 스트라이크 추가 점수를 반환합니다.

imcool2551 committed:
feat+test: FinalFrame

10 프레임을 나타내는 FinalFrame 클래스입니다.

imcool2551 committed:
feat+test: NormalFrame

1~9프레임을 나타내는 NormalFrame 클래스입니다.

imcool2551 committed:
feat+test: Frames

1~10 프레임중 점수 합산이 가능한 프레임들만 모아서 점수를 합산한 뒤 리스트로 반환하는 메서드를 추가하고 테스트했습니다.

imcool2551 committed:
feat: OutputView

현재 점수 합산이 가능한 프레임들의 점수를 출력하는 기능을 추가했습니다.

yongdae commented on README.md:
```diff
@@ -2,49 +2,26 @@
+* 스페어는 다음 한 번의 투구에 쓰러뜨린 핀의 개수만큼 보너스 점수를 추가한다.
+* 10프레임은 스트라이크든 아니든 기본적으로 두 번을 투구한다.
+* 10프레임이 스트라이크/스페어라면 한 번을 더 투구한다.
+* 10프레임의 점수 계산 방식도 일반 프레임과 동일하다.
```
<img width="499" alt="스크린샷 2022-05-16 오후 9 49 46" src="https://user-images.githubusercontent.com/4317804/168596208-14e96747-b629-459d-8e78-8c1335cb785f.png">

스트라이크가 누적되는 경우에는 3번째에 점수가 계산되어야해요.

<img width="625" alt="스크린샷 2022-05-16 오후 9 50 02" src="https://user-images.githubusercontent.com/4317804/168596408-fee03315-8a4c-4940-92a7-6a1d94f45ac5.png">

볼링 점수 계산에 대한 정확한 내용은 [링크](https://www.bowlinggenius.com/)를 통해 확인해보셔도 좋을 것 같아요. :)

yongdae commented on src/main/java/bowling/domain/frame/NormalFrame.java:
```diff
@@ -31,15 +35,55 @@ public void addPin(int pinNo) {
+            return false;
+        }
+
+        FrameResult result = pinNumbers.getResult();
```
FrameResult에 상태를 알 수 있는 메서드를 추가한다면 
canGetScore()에서 직접 프레임이 종료되었는지 여부를 판단하지 않아도 되지 않을까요?

    boolean isFinish();

yongdae commented on src/main/java/bowling/domain/frameresult/FrameResult.java:
```diff
@@ -0,0 +1,6 @@
+package bowling.domain.frameresult;
+
+public interface FrameResult {
```
피드백을 잘 반영해주셨네요. 💯

yongdae commented on src/main/java/bowling/domain/frameresult/FrameResult.java:
```diff
@@ -0,0 +1,6 @@
+
+public interface FrameResult {
+
+    int getScoreWithBonus(int bonus);
```
<img width="755" alt="스크린샷 2022-05-16 오후 9 52 57" src="https://user-images.githubusercontent.com/4317804/168596754-8a7d0337-5f3e-43f6-99a3-f5f3a96c1154.png">

점수를 의미하는 객체를 추가하고 FrameResult의 getScoreWithBonus()에서
그 객체를 활용해서 점수를 계산하게 해본다면 어떨까요? 🤔 

메서드의 이름도 calculateScore() 같은 `계산`의 의미를 포함하는 이름으로 바꾼다면 더 좋을 것 같아요.

yongdae reviewed (changes requested):
볼링 점수 계산이 쉽지 않은 부분인데 잘 만들어주셨네요. 👍 

의견주신 부분을 포함해 점수 계산과 관련 된 부분에 대해 의견을 남겨 놓았습니다.
확인하시고 다시 리뷰 요청 부탁드릴께요! 🔥

imcool2551 committed:
refactor: Frame

Frame 구현체가 결과를 도출할 수 있는지 나타내는 메서드 이름을 변경했습니다. 동작방식은 그대로입니다.
이전 프레임을 위한 스트라이크 추가 점수를 반환하는 메서드가 Optional을 반환하도록 했습니다. 아직 투구가 채워지지 않아서 추가 점수를 계산할 수 없는 경우 빈 Optional을 반환합니다.

imcool2551 committed:
refactor+test: FinalFrame

마지막 프레임의 경우 투구를 2번해야 이전 프레임을 위한 스트라이크 추가 점수를 반환할 수 있도록 했습니다.

imcool2551 committed:
refactor+test: NormalFrame

일반 프레임의 경우 스트라이크 추가 점수는 다음 경우가 있습니다.
1. 현재 프레임의 투구가 끝나지 않으면 계산할 수 없으므로 빈 Optional 반환
2. 현재 프레임의 투구가 끝나도 현재 프레임이 스트라이크라면 다음 프레임에 한 번더 투구해야함.
   2.1 다음 프레임이 없다면 빈 Optional 반환
   2.2 다음 프레임이 있다면 10점에 다음 프레임의 Spare 보너스를 합산해서 반환
3. 현재 프레임의 투구가 끝나고 현재 프레임이 스트라이크가 아니면 바로 반환.

imcool2551 committed:
refactor: Frames

점수를 출력할 때 계산된 점수를 바로 반환받는 대신 결과가 나온 프레임을 1차 선별하고 그 중에서 점수 합산이 가능한 프레임을 2차 선별해서 점수를 출력했습니다.

imcool2551 committed:
refactor: FrameResult

프레임 결과 인터페이스를 변경했습니다.
1. caculateScore(): 프레임 점수 합산
2. isCalculated(): 프레임 점수를 합산할 수 있는지 반환. (스트라이크/스페어의 경우 보너스 점수를 추가해야 참 반환)
3. addBonus(int bonus): 스트라이크/스페어의 경우 보너스 점수 추가

imcool2551 committed:
refactor: FrameResult Implementations

Miss, Spare, Strike를 인터페이스에 맞게 구현했습니다. Spare, Strike의  구현은 똑같습니다.

imcool2551 committed:
refactor: minor format changes

yongdae commented on src/main/java/bowling/domain/frame/Frame.java:
```diff
@@ -1,12 +1,22 @@
+
+    FrameResult getResult();
+
+    int spareBonusForPreviousFrame();
+
+    Optional<Integer> strikeBonusForPreviousFrame();
```
점수 계산을 위해 특정 상황(spare, strike)에 따라 메소드를 구분해서 호출하지 않고
점수를 계산 할 수 있는 방법은 없을까요?

이전에 드린 [의견](https://github.com/next-step/java-bowling/pull/803#discussion_r873705955)과 맞물려 있는 부분이에요.
점수를 표현하는 객체를 추가해서 점수를 계산과 관련 된 기능을 넣어보면 어떨까요? 🤔

yongdae commented on src/main/java/bowling/domain/frameresult/Spare.java:
```diff
@@ -0,0 +1,26 @@
+
+public class Spare implements FrameResult {
+
+    private Integer bonus;
```
앞서 의견 드린 부분과 동일한 부분인 것 같아요.
Spare, Strike가 bonus를 가지지 않고 점수 계산을 할 수 있는 방법은 없을까요?

yongdae commented on src/main/java/bowling/domain/frame/NormalFrame.java:
```diff
@@ -31,15 +38,48 @@ public void addPin(int pinNo) {
+            return Optional.empty();
+        }
+        if (pinNumbers.getResult() instanceof Strike) {
+            return nextFrame == null
```
3항 연산자는 else를 사용하는 것과 같습니다.
`규칙 2: else 예약어를 쓰지 않는다.` 이 규칙을 적용해보면 좋을 것 같아요.

yongdae reviewed (changes requested):
피드백을 잘 반영해주셨네요. 👍 
다만 점수와 관련 된 부분은 아직 반영이 되지 않은 것 같아요.

만약 이 부분에 대해 다른 의견이 있으시다면
코멘트나 DM을 통해 이야기 해주시면 좋을 것 같습니다. 😄 

확인하시고 다시 리뷰 요청 부탁드릴께요! 🙇

imcool2551 committed:
feat: Bonus

Spare, Strike에 해당하는 추가 점수를 담아두는 클래스입니다.

imcool2551 committed:
refactor: FrameResult

FrameResult가 Bonus를 통해 점수를 계산하도록 변경했습니다.

imcool2551 committed:
refactor: FrameResult Implementaions and tests

FrameResult의 구현체를 인터페이스에 맞게 수정하고, 테스트 그에 맞게 수정했습니다.

imcool2551 committed:
refactor: Frame

Frame인터페이스가 FrameResult를 직접반환하는 대신 계산된 점수를 직접 반환하도록 변경했습니다.

imcool2551 committed:
refactor: FinalFrame and test

getResult() 구현에서 Frame의 결과와 상관 없이 보너스를 추가해서 결과에 따른 분기처리를 제거했습니다.
FinalPinNumbers를 참조해서 보너스를 추가할 때 IndexOutOfBounds 예외 때문에 Miss는 어쩔 수 없이 분기처리를 했습니다.

imcool2551 committed:
refactor: NormalFrame and test

FinalFrame과 마찬가지로 getResult()에서 보너스 점수를 항상 추가해서 프레임의 결과에 따른 분기처리를 삭제했습니다.

imcool2551 committed:
refactor: Frames

Frames 클래스를 Frame인터페이스 수정에 맞춰 변경했습니다. Optional의 존재 여부를 검사해서 점수 집계가 완료된 프레임의 점수만 모아서 반환합니다.

yongdae commented on src/main/java/bowling/domain/frameresult/FrameResult.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface FrameResult {
+
+    Optional<Integer> calculateScore(Bonus bonus);
```
[이전과 같은 의견](https://github.com/next-step/java-bowling/pull/803#discussion_r873705955)이에요.
`보너스`가 아닌 `점수`를 표현하는 객체를 추가해서 점수를 계산해보면 좋을 것 같아요. :)

yongdae commented on src/main/java/bowling/domain/frame/Frame.java:
```diff
@@ -1,12 +1,20 @@
+
+    Optional<Integer> getResult();
+
+    int spareBonusForPreviousFrame();
 
+    Optional<Integer> strikeBonusForPreviousFrame();
```
이 부분도 이전과 동일한 [의견](https://github.com/next-step/java-bowling/pull/803#discussion_r875010788)이에요.
Frame이 스페어, 스트라이크 그리고 보너스까지 메소드를 통해 분기 하지 않고 점수를 계산 할 수 있게 해본다면 어떨까요?

앞서 의견드린 `점수` 객체가 추가 된다면 지금과는 다른 구조로 점수를 계산 할 수 있을 것 같아요.

yongdae reviewed (changes requested):
DM에서 나누었던 대화처럼 이전에 의견드린 부분을 반영해보시고
다시 리뷰 요청 부탁드리겠습니다. 🙇 

고민이나 의견이 있으시다면 언제든 이야기 해주세요!! :)

imcool2551 committed:
save before restart

imcool2551 committed:
START OVER

재설계를 위한 파일 삭제

imcool2551 committed:
feat+test: FrameNo

Frame의 번호를 나타내는 클래스입니다.

imcool2551 committed:
feat+test: Pin

투구 한 번에 쓰러진 핀의 개수를 나타내는 클래스입니다.

imcool2551 committed:
feat: Frame

프레임을 나타내는 인터페이스입니다. 구현체로는 Normalframe, FinalFrame 이 있습니다.
1) int number(): 프레임의 번호 반환(1~10)
2) Frame bowl(Pin no): 투구 추가. 투구 후에도 프레임이 진행중이라면 자신 반환. 프레임이 투구로 인해 종료되면 다음 프레임 반환
3) Score score(): 점수를 나타내는 객체 반환
4) boolean finished(): 프레임 종료 여부 반환
5) Score additionalScore(Score beforeScore): 이전 프레임의 점수 객체를 받아서 점수를 추가한 점수 객체 반환
6) String expression(): 프레임의 문자열 표현 반환. ex) X|X|X

imcool2551 committed:
feat+test: NormalFrame

1~9 번 프레임을 나타내는 클래스입니다.
다음 프레임을 생성해서 필드로 설정합니다.
bowl() 구현에서 9번프레임이 종료되면 FinalFrame을 반환합니다.

imcool2551 committed:
feat+test: FinalFrame

10번프레임을 나타내는 클래스입니다.
10번프레임은 추가 투구를 할 수 있기 때문에 추가 투구한 Pin들을 관리하는 FinalExtraPins 일급 컬렉션을 필드로 가집니다.

imcool2551 committed:
feat+test: FinalExtraPins

마지막 프레임이 SPARE/STRIKE 일 때 추가 투구한 Pin들을 리스트로 관리합니다.
SPARE의 경우 추가 투구를 1번만 추가할 수 있고, STRIKE는 2번 추가할 수 있습니다.
FinalFrame에서 추가 점수를 구하는 additionalScore()도 SPARE/STRIKE에 따라 각각 spareScore(), strikeScore() 로 위임합니다.
FinalFrame의 문자열 표현을 구하는 expression()도 마찬가지로 이 클래스의 expression()으로 위임해서 처리합니다.

imcool2551 committed:
feat+test: Frames

Main에서 사용하는 Frames클래스입니다. 1~10번 프레임을 리스트로 관리합니다.
게임 종료 여부는 마지막 FinalFrame이 종료됐는지 여부로 판단하고, 종료전 까지 투구를 할 수 있습니다.
View에서 사용할 Frame들의 점수와 문자열 표현을 각각 리스트로 반환합니다.

imcool2551 committed:
feat+test: Player

imcool2551 committed:
feat: State

프레임의 상태를 나타내는 인터페이스입니다. 모든 프레임은 State를 가지고 State에 따라서 동작이 변경됩니다.
1) State bowl(Pin no): 투구를하고 변화된 상태를 반환합니다.
2) Score score(): 점수 객체를 반환합니다.
3) boolean finished(): 현재 상태가 종료 상태인지 반환합니다.
4) Score additionalScore(Score beforeScore): 추가 점수를 추가하기 전의 점수 객체를 받아서, 추가 점수를 추가한 점수 객체를 반환합니다.
5) String expression(): 상태의 문자열 표현을 반환합니다.

imcool2551 committed:
feat+test: Running State and its implementations (Ready, FirstBowl)

아직 종료되지 않은 상태를 표현하는 Ready, FirstBowl 클래스입니다.
모든 프레임은 아무 투구도 하지 않은 Ready 상태로 시작하고 첫 투구를 하면 FirstBowl 또는 Strike 상태가 됩니다. (Strike는 프레임이 종료되기 때문에 Running 의 구현체가 아닙니다)
FirstBowl 상태에서는 두 번째 투구에 쓰러뜨린 핀의 개수에 따라 Miss, Spare로 변화합니다.

imcool2551 committed:
feat+test: Finished State and it's implementations (Miss, Spare, Strike)

종료된 상태를 나타내는 Finished와 그 구현체들입니다.

imcool2551 committed:
feat+test: Score

점수를 나타내는 클래스입니다.

score는 누적 점수, left는 남은 투구 개수를 나타냅니다.
투구 개수가 남아있다면 계산된 점수를 구할 수 없습니다.
addedScore()를 호출할 때 마다 투구 개수를 감소시킨 대신 누적 점수를 증가한 새로운 점수 객체를 반환합니다.

imcool2551 committed:
feat: UI+Main

yongdae commented on src/main/java/bowling/Main.java:
```diff
@@ -1,23 +1,20 @@
-
 public class Main {
 
     public static void main(String[] args) {
```
<img width="668" alt="스크린샷 2022-05-22 오전 10 40 37" src="https://user-images.githubusercontent.com/4317804/169674735-816e63da-2527-4774-bc0c-70280714904a.png">

👍

yongdae commented on src/main/java/bowling/domain/state/State.java:
```diff
@@ -0,0 +1,17 @@
+import bowling.domain.pin.Pin;
+import bowling.domain.Score;
+
+public interface State {
```
프레임의 상태를 인터페이스로 잘 분리해주셨네요. 
특히 State의 구현체들은 정말 깔끔하게 정리해주셨다고 생각해요. 💯

yongdae commented on src/main/java/bowling/domain/state/Finished.java:
```diff
@@ -0,0 +1,16 @@
+
+import bowling.domain.pin.Pin;
+
+public abstract class Finished implements State {
```
간결하고 멋지게 종료 된 상태를 멋지게 표현해주셨네요. 👍

yongdae commented on src/main/java/bowling/domain/Score.java:
```diff
@@ -0,0 +1,41 @@
+
+import bowling.domain.pin.Pin;
+
+public class Score {
```
남아 있는 계산 시도 횟수를 포함시켜 볼링 게임에서의 `점수`를 잘 설계 해주셨네요. :)

yongdae commented on src/test/java/bowling/domain/ScoreTest.java:
```diff
@@ -0,0 +1,35 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class ScoreTest {
```
점수 객체에 대한 테스트도 잘 만들어주셨네요!

yongdae commented on src/test/java/bowling/domain/ScoreTest.java:
```diff
@@ -0,0 +1,35 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+class ScoreTest {
+
```
@Test
    public void addedScore_X81() {
        Score score = new Score(10, 2);
        Score last = score.addedScore(Pin.of(8))
                .addedScore(Pin.of(1));

        assertThat(last.getScore()).isEqualTo(19);
    }

    @Test
    public void addedScore_XX9() {
        Score score = new Score(10, 2);
        Score last = score.addedScore(Pin.of(10))
                .addedScore(Pin.of(9));
        
        assertThat(last.getScore()).isEqualTo(29);
    }

이런 테스트 코드를 추가해보면 어떨까요?
FinalFrame에서의 Strike, Spare 점수를 테스트를 해볼 수 있을 것 같아요. :)

yongdae commented on src/main/java/bowling/domain/frame/FinalFrame.java:
```diff
@@ -1,64 +1,82 @@
+    private static final int FINAL_FRAME_NO = 10;
 
-public class FinalFrame implements Frame {
+    private State state;
```
FinalFrame의 상태를 State를 1개 이상을 가지게 해서 정의해 본다면
FinalExtraPins를 사용하지 않고도 마지막 프레임을 장식 해볼 수 있을 것 같아요!

yongdae reviewed (approved):
기존의 설계를 버리고 재설계하는게 쉽지 않으셨을텐데
새로운 구조로 깔끔하게 설계를 해주셨네요. 💯 

진행해주신 내용에 대해서는 소소하지만 몇 가지 의견을 남겨놓았습니다.
이번 피드백은 다음 단계에서 같이 반영을 부탁드릴께요 ~ 😄 

자 그럼 이제 마지막 단계로 넘어가도록 하겠습니다.
수고 많으셨습니다. 🔥🔥
